src/cargo/core/workspace.rs

```diff
@@ -1669,31 +1685,35 @@ pub fn resolve_relative_path(
     }
 }
 
-fn parse_manifest(manifest_path: &Path, config: &Config) -> CargoResult<EitherManifest> {
-    let key = manifest_path.parent().unwrap();
-    let source_id = SourceId::for_path(key)?;
-    let (manifest, _nested_paths) = read_manifest(manifest_path, source_id, config)?;
-    Ok(manifest)
+/// Finds the path of the root of the workspace.
+pub fn find_workspace_root(manifest_path: &Path, config: &Config) -> CargoResult<Option<PathBuf>> {
+    // FIXME(ehuss): Loading and parsing manifests just to find the root seems
+    // very inefficient. I think this should be reconsidered.
```

Comment:
The performance of this was tested before merging [in this PR](https://github.com/rust-lang/cargo/pull/10568#issue-1204940688). There is some hit to the performance but not enough that it was a concern at the time. There was hope that any concerns about performance would be found by people testing the feature but none have come up so far.

If it is something that needs to be worked on the easiest fix would be to record workspace root paths once they have been found. Then each time a workspace root is needed you check if an already existing workspace root is the correct one for the current manifest. The idea is similar to what I described in [this comment](https://github.com/rust-lang/cargo/pull/9684#issuecomment-882635097).

Reply:
I just did a quick test of using workspace inheritance in the rust workspace.  It increased the startup time from 358.8 ms to 449.1 ms. That's a pretty substantial increase.  I didn't do any profiling to see if this particular thing is the culprit, but I suspect it is.

Reply:
If you don't mind sharing what you did for the test, I would like to take a look at it. That is much higher than I have seen but I think that is because of the nesting in the rust repo. 

I think caching paths is the best way to eliminate most of that but I don't know for certain. I will take a look at it and see what I can do. I don't want to clutter this PR so if a separate issue is better for further discussion I would be happy to open one.

Reply:
1. Extract this file in `benches/workspaces`
   [rust2.zip](https://github.com/rust-lang/cargo/files/8881641/rust2.zip)

   This is just a quick hack that adds inheritance to a bunch of packages, but I didn't really put much thought into setting it up.

2. Extract `rust.tgz` in that directory, too.

3. Prime the target directory with a cache of `rustc` info. In `rust` and `rust2`, run: `cargo c -p linkchecker`. Otherwise it would be measuring `rustc` overhead.

4. Test out some command that spends most of its time loading the workspace. There aren't very many good candidates, but `cargo metadata` or `cargo tree` can work. I use hyperfine for quick measurements. For example:

   ```bash
   hyperfine "cd rust; $(rustup which cargo) metadata" "cd rust2; $(rustup which cargo) metadata"
   ```

We really should add workspace initialization benchmarks. If you want to work on that, that would be wonderful. I think that would just require adding a new benchmark to `benches/benchsuite/benches` that essentially just calls `cargo::core::Workspace::new`. Might have to factor out the common logic from `resolve.rs` that sets up the workspaces.